File: client-android/app/src/main/java/app/serenada/android/push/PushSubscriptionManager.kt
package app.serenada.android.push

import android.content.Context
import android.util.Log
import app.serenada.android.data.SettingsStore
import app.serenada.android.network.ApiClient
import app.serenada.android.network.PushSubscribeRequest
import com.google.firebase.messaging.FirebaseMessaging
import java.util.Locale

class PushSubscriptionManager(
    context: Context,
    private val apiClient: ApiClient,
    private val settingsStore: SettingsStore,
    private val pushKeyStore: PushKeyStore = PushKeyStore()
) {
    private val appContext = context.applicationContext

    fun cachedEndpoint(): String? = settingsStore.pushEndpoint?.trim()?.ifBlank { null }

    fun refreshPushEndpoint(onResult: (String?) -> Unit) {
        val cached = cachedEndpoint()
        if (!cached.isNullOrBlank()) {
            onResult(cached)
            return
        }

        if (!FirebasePushInitializer.ensureInitialized(appContext)) {
            onResult(null)
            return
        }

        runCatching {
            FirebaseMessaging.getInstance().token
                .addOnCompleteListener { task ->
                    if (!task.isSuccessful) {
                        Log.w("Push", "Failed to fetch FCM token", task.exception)
                        onResult(null)
                        return@addOnCompleteListener
                    }
                    val token = task.result?.trim().orEmpty()
                    if (token.isBlank()) {
                        onResult(null)
                        return@addOnCompleteListener
                    }
                    settingsStore.pushEndpoint = token
                    onResult(token)
                }
        }.onFailure {
            Log.w("Push", "Unable to request FCM token", it)
            onResult(null)
        }
    }

    fun subscribeRoom(roomId: String, host: String) {
        val normalizedRoomId = roomId.trim()
        if (normalizedRoomId.isBlank()) return

        refreshPushEndpoint { endpoint ->
            val token = endpoint?.trim().orEmpty()
            if (token.isBlank()) {
                Log.w("Push", "Skipping room subscription: missing FCM token")
                return@refreshPushEndpoint
            }

            val publicJwk = pushKeyStore.getPublicJwk()
            val request = PushSubscribeRequest(
                transport = PUSH_TRANSPORT,
                endpoint = token,
                locale = Locale.getDefault().toLanguageTag(),
                encPublicKey = publicJwk
            )

            apiClient.subscribePush(host, normalizedRoomId, request) { result ->
                result.onSuccess {
                    Log.d("Push", "Subscribed Android push for room $normalizedRoomId")
                }.onFailure {
                    Log.w("Push", "Failed to subscribe Android push for room $normalizedRoomId", it)
                }
            }
        }
    }

    companion object {
        const val PUSH_TRANSPORT = "fcm"
    }
}

File: client-ios/Sources/Core/Push/PushSubscriptionManager.swift
import Foundation
#if canImport(FirebaseCore)
import FirebaseCore
#endif
#if canImport(FirebaseMessaging)
import FirebaseMessaging
#endif

@MainActor
final class PushSubscriptionManager {
    static let pushTransport = "fcm"

    private let apiClient: APIClient
    private let settingsStore: SettingsStore
    private let pushKeyStore: PushKeyStore

    init(
        apiClient: APIClient,
        settingsStore: SettingsStore,
        pushKeyStore: PushKeyStore = PushKeyStore()
    ) {
        self.apiClient = apiClient
        self.settingsStore = settingsStore
        self.pushKeyStore = pushKeyStore
    }

    func cachedEndpoint() -> String? {
        let cached = settingsStore.pushEndpoint?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let cached, !cached.isEmpty else { return nil }
        return cached
    }

    func updateCachedEndpoint(_ endpoint: String?) {
        settingsStore.pushEndpoint = endpoint
    }

    func refreshPushEndpoint() async -> String? {
        if let cached = cachedEndpoint() {
            log("Using cached push endpoint for room subscription")
            return cached
        }

        #if canImport(FirebaseMessaging)
        if FirebaseApp.app() != nil, let token = await fetchFirebaseToken() {
            settingsStore.pushEndpoint = token
            log("Using Firebase Messaging token for push subscriptions")
            return token
        }
        #endif

        log("Push endpoint unavailable for room subscription")
        return nil
    }

    func subscribeRoom(roomId: String, host: String) {
        let normalizedRoomId = roomId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedRoomId.isEmpty else { return }

        Task { [weak self] in
            guard let self else { return }
            guard let endpoint = await self.refreshPushEndpoint() else {
                self.log("Skipping push subscription for room \(normalizedRoomId): missing push endpoint")
                return
            }
            guard let encPublicKeyRaw = self.pushKeyStore.getPublicJwk() else {
                self.log("Skipping push subscription for room \(normalizedRoomId): missing encryption public key")
                return
            }
            let encPublicKey = PushRecipientPublicKey(
                kty: encPublicKeyRaw.kty,
                crv: encPublicKeyRaw.crv,
                x: encPublicKeyRaw.x,
                y: encPublicKeyRaw.y
            )

            self.log("Attempting push subscribe for room \(normalizedRoomId)")
            let request = PushSubscribeRequest(
                transport: Self.pushTransport,
                endpoint: endpoint,
                locale: Locale.current.languageTagCompatible(),
                encPublicKey: encPublicKey,
                auth: nil,
                p256dh: nil
            )

            do {
                try await self.apiClient.subscribePush(host: host, roomId: normalizedRoomId, request: request)
                self.log("Subscribed room \(normalizedRoomId) for push")
            } catch {
                self.log("Push subscribe failed for room \(normalizedRoomId): \(error.localizedDescription)")
            }
        }
    }

    #if canImport(FirebaseMessaging)
    private func fetchFirebaseToken() async -> String? {
        await withCheckedContinuation { continuation in
            Messaging.messaging().token { token, error in
                if let error {
                    PushDiagnostics.append("Failed to fetch Firebase Messaging token: \(error.localizedDescription)")
                }
                let clean = token?.trimmingCharacters(in: .whitespacesAndNewlines)
                continuation.resume(returning: (clean?.isEmpty == false) ? clean : nil)
            }
        }
    }
    #endif

    private func log(_ message: String) {
        PushDiagnostics.append(message)
    }
}

private extension Locale {
    func languageTagCompatible() -> String {
        if #available(iOS 16.0, *) {
            return self.language.languageCode?.identifier ?? identifier.replacingOccurrences(of: "_", with: "-")
        }
        return identifier.replacingOccurrences(of: "_", with: "-")
    }
}
